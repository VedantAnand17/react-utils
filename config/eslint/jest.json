{
  "env": {
    "browser": true,
    "jest": true
  },
  "plugins": [
    "jest"
  ],
  "rules": {
    "@typescript-eslint/no-implied-eval": 0,
    "global-require": 0,
    "import/no-dynamic-require": 0,
    "import/no-extraneous-dependencies": 0,
    "no-console": 0,
    "no-new-func": 0,
    "react/require-default-props": 0
  }
}