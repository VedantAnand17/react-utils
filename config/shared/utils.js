/**
 * Misc utilities reusable in different JS-format config files.
 */
/* eslint-disable import/no-extraneous-dependencies */

const fs = require('fs');
const path = require('path');
const cssesc = require('cssesc');

const { interpolateName } = require('loader-utils');

/**
 * Normalizes file path to OS-independent format (adopted from css-loader).
 * @ignore
 * @param {string} file
 * @return {string}
 */
function normalizePath(file) {
  return path.sep === '\\' ? file.replace(/\\/g, '/') : file;
}

// eslint-disable-next-line no-control-regex
const filenameReservedRegex = /[<>:"/\\|?*]/g;
// eslint-disable-next-line no-control-regex
const reControlChars = /[\u0000-\u001f\u0080-\u009f]/g;

function escapeLocalident(localident) {
  return cssesc(
    localident
      // For `[hash]` placeholder
      .replace(/^((-?[0-9])|--)/, '_$1')
      .replace(filenameReservedRegex, '-')
      .replace(reControlChars, '-')
      .replace(/\./g, '-'),
    { isIdentifier: true },
  );
}

/**
 * Returns the name of package containing the folder; i.e. it recursively looks
 * up from the folder for the closest package.json file, and returns the name in
 * that file. It also caches the results from previously fisited folders.
 * @ignore
 * @param {string} folder
 * @return {string}
 */
function getPackageInfo(folder) {
  let res = getPackageInfo.cache[folder];
  if (!res) {
    const p = path.resolve(folder, 'package.json');
    /* eslint-disable global-require, import/no-dynamic-require */
    res = fs.existsSync(p) ? {
      name: require(p).name,
      root: folder,
    } : getPackageInfo(path.resolve(folder, '..'));
    /* eslint-enable global-require, import/no-dynamic-require */
    getPackageInfo.cache[folder] = res;
  }
  return res;
}
getPackageInfo.cache = {};

function getLocalIdent(
  { resourcePath },
  localIdentName,
  localName,
  options = {},
) {
  const packageInfo = getPackageInfo(path.dirname(resourcePath));
  const request = normalizePath(path.relative(packageInfo.root, resourcePath));
  return interpolateName({
    resourcePath,
  }, localIdentName, {
    ...options,
    content: `${packageInfo.name + request}\x00${localName}`,
    context: packageInfo.root,
  }).replace(/\[package\]/gi, packageInfo.name)
    .replace(/\[local\]/gi, localName)
    .replace(/@/g, '-');
}

function generateScopedNameDev(localName, assetPath) {
  // Mind: "css-loader" additionally escapes identifiers returned by
  // getLocalIdent(); the babel plugins - don't, thus we need to escape
  // ourselves.
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[package]___[path][name]___[local]___[hash:base64:6]',
      localName,
      {},
    ),
  );
}

function generateScopedNameProd(localName, assetPath) {
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[hash:base64:6]',
      localName,
      {},
    ),
  );
}

module.exports = {
  generateScopedNameDev,
  generateScopedNameProd,
  getLocalIdent,
};
