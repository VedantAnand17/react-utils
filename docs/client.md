# Client

**TODO: To be revised! Does not seem to match the actual implementation!**

Client-side intialization code.

- **Why?** &mdash; For the same reasons we have the standard
  [server setup](./server.md).

This module exports a single **`Launch(options)`** function that starts the app
at the client side. It renders the app, wraps it with
[`react-router`](https://www.npmjs.com/package/react-router)'s `<BrowserRouter>`,
and Redux store `<Provider>`, if necessary; then injects it into the page HTML
template generated by the server. In the case of use with development server,
it also takes care about HMR (Hot Module Reloading) setup at the clients-side.

**`options`** should be a JS object, with the following fields:
- **`Application`** &mdash; *Function* &mdash; The root ReactJS component of
  the app to be rendered.

  **Arguments**
  - **`initialState`** &mdash; *Object* &mdash; Initial global state, injected by
    the server (will be `undefined`, if it is not generated by the current
    server setup).

  **Returns**
  - *Promise* that resolves to the created Redux store.
