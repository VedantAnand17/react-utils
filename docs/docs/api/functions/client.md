# client()
```jsx
import { client } from '@dr.pogodin/react-utils';

client(Application, options);
```
Initializes and launches ReactJS `Application` at the client-side.

Namely, it wraps the app with [BrowserRouter] and [GlobalStateProvider],
then hydrates it into DOM element with **react-view** ID, assuming
[the standard SSR setup] was used to generate the initial HTML markup of the page.

The `dontHydrate` option allows to opt for a pure client-side rendering, instead
of the hydration.

## Arguments
- `Application` - **React.Component** - The root applciation component.
- `options` - **object** - Optional. Additional settings:
  - `dontHydrate` - **boolean** - By default, the app is hydrated into DOM
    element found by **react-view** ID, using React's [hydrateRoot()] functionality,
    and  assuming [the standard SSR setup] was used to generate the initial HTML markup of the page.

    With `dontHydrate` flag set **true** React's [createRoot()] function is used
    instead to perform a clean client-side rendering into the DOM element with
    **react-view** ID.

<!-- links -->
[BrowserRouter]: https://reactrouter.com/docs/en/v6/api#browserrouter
[GlobalStateProvider]: https://dr.pogodin.studio/docs/react-global-state/docs/api/components/globalstateprovider
[the standard SSR setup]: /docs/api/functions/server
[createRoot()]: https://reactjs.org/docs/react-dom-client.html#createroot
[hydrateRoot()]: https://reactjs.org/docs/react-dom-client.html#hydrateroot