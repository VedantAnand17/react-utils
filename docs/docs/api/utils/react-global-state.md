# React Global State
```jsx
import {
  getGlobalState,
  getSsrContext,
  GlobalStateProvider,
  useAsyncCollection,
  useAsyncData,
  useGlobalState,
} from '@dr.pogodin/react-utils';
```

For convenience, **react-utils** re-exports components and hooks provided by
[@dr.pogodin/react-global-state](https://dr.pogodin.studio/docs/react-global-state/index.html)
library, which  is a state of the art library for the global state and
asynchronous data managment in React applications, powered by hooks and Context
API. It is simple, efficient, and with a full server-side rendering (SSR)
support.

The standard [client] and [server] setups provided by **react-utils**
automatically wrap the app into `<GlobalStateProvider>`, also the [server] wires
**react-global-state** to **react-utils** SSR setup, and also exposes the actual
incoming HTTP request object via SSR context, attaching it as `req` field to
the context.

[client]: /docs/api/functions/client
[server]: /docs/api/functions/server
