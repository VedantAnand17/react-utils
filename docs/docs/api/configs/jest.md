# Jest
Default configuration for [Jest] testing framework.

:::caution Beware
It does strong assumptions about the host project's code structure: the host
[Jest] configuration in the `/config/jest` folder of host code; the source code
in `/src` folder; build outputs in `/build` folder; _etc._

For a project with different structure it might be easier to compose
the appropriate Jest config from scratch.
:::

To use this default configuration create two files in the `/config/jest` folder
of host project:

```js title="/config/jest/config.js"
const config = require('@dr.pogodin/react-utils/config/jest/default');

// You can tune config as necessary here.

module.exports = config;
```

```js title="/config/jest/setup.js"
import '@dr.pogodin/react-utils/config/jest/setup';
```

After that point to that host config file from the Jest execution command,
_e.g._:
```bash
jest --config config/jest/config.js
```

[Jest]: https://jestjs.io
