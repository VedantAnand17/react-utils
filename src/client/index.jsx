// Initialization of client-side code.
/* global document, window */

import { GlobalStateProvider } from '@dr.pogodin/react-global-state';

import { hydrateRoot } from 'react-dom/client';
import { BrowserRouter } from 'react-router-dom';

/**
 * Prepares and launches the app at client side.
 * @param {object} Application Root application component
 */
export default function Launch(Application) {
  hydrateRoot(
    document.getElementById('react-view'),
    <GlobalStateProvider initialState={window.ISTATE}>
      <BrowserRouter>
        <Application />
      </BrowserRouter>
    </GlobalStateProvider>,
  );
}
