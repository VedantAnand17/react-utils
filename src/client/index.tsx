// Initialization of client-side code.
/* global document */

import { GlobalStateProvider } from '@dr.pogodin/react-global-state';

import { createRoot, hydrateRoot } from 'react-dom/client';
import { BrowserRouter } from 'react-router-dom';

import getInj from './getInj';

/**
 * Prepares and launches the app at client side.
 * @param {object} Application Root application component
 * @param {object} [options={}] Optional. Additional settings.
 */
export default function Launch(Application, options = {}) {
  const container = document.getElementById('react-view');
  const scene = (
    <GlobalStateProvider initialState={getInj().ISTATE}>
      <BrowserRouter>
        <Application />
      </BrowserRouter>
    </GlobalStateProvider>
  );

  if (options.dontHydrate) {
    const root = createRoot(container);
    root.render(scene);
  } else hydrateRoot(container, scene);
}
