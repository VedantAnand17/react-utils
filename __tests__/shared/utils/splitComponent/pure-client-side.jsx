/** @jest-environment jsdom */

// This tests pure client-side rendering from scratch.

import { GlobalStateProvider } from '@dr.pogodin/react-global-state';
import pretty from 'pretty';
import { BrowserRouter } from 'react-router-dom';
import { act } from 'react-test-renderer';
import { mockClientSide, mount } from 'utils/jest';

test('Pure client-side render', async () => {
  mockClientSide();

  window.TRU_BUILD_INFO = {
    publicPath: '/test/public/path',
  };

  require('client/init');

  window.CHUNK_GROUPS = {
    main: [
      'main-1511941200000.css',
      'main-1511941200000.js',
    ],
    'sample-component-1': [
      'sample-component-1-1511941200000.css',
    ],
    'sample-component-2': [
      'sample-component-2-1511941200000.js',
      'sample-component-2-1511941200000.css',
    ],
    'sample-component-3': [
      'sample-component-3-1511941200000.js',
      'sample-component-3-1511941200000.css',
    ],
  };

  const SampleCodeSplit = require('./__assets__/SampleCodeSplit').default;

  // Note: <BrowserRouter> is needed here because we don't use the client-side
  // initializer here, which would wrap with it automatically.
  const scene = mount(
    <BrowserRouter>
      <GlobalStateProvider>
        <SampleCodeSplit />
      </GlobalStateProvider>
    </BrowserRouter>,
  );

  // Note: this is a working way to wrap async Suspense resolution in component
  // tree within act().
  await act(() => new Promise(process.nextTick));

  // TODO: The current head snapshot is not correct, as misses the actual
  // <link> elements we expect to have.
  expect(pretty(document.head.innerHTML)).toMatchSnapshot();

  expect(pretty(scene.innerHTML)).toMatchSnapshot();

  await act(() => {
    document.head.querySelectorAll('link').forEach((link) => {
      link.onload();
    });
    return new Promise(process.nextTick);
  });

  expect(pretty(scene.innerHTML)).toMatchSnapshot();
});
