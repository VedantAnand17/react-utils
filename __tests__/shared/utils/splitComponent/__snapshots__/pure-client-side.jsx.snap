// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pure client-side render 1`] = `
"<link href=\\"/test/public/path/sample-component-1-1511941200000.css\\" rel=\\"stylesheet\\">
<link href=\\"/test/public/path/sample-component-2-1511941200000.css\\" rel=\\"stylesheet\\">
<link href=\\"/test/public/path/sample-component-3-1511941200000.css\\" rel=\\"stylesheet\\">"
`;

exports[`Pure client-side render 2`] = `
"<h1>SampleCodeSplit</h1>
<div>testValue</div>
<div>PLACEHOLDER!</div>"
`;

exports[`Pure client-side render 3`] = `
"<h1>SampleCodeSplit</h1>
<div>testValue</div>
<div>
  <h1>Sample Component</h1>
  <div>testValue</div>
  <h3>Prop</h3>
  <div></div>
  <h3>Children</h3>
  <div></div>
</div>
<div>
  <h1>Sample Component</h1>
  <div>testValue</div>
  <h3>Prop</h3>
  <div>test</div>
  <h3>Children</h3>
  <div>Test</div>
</div>
<div>
  <h1>Sample Component</h1>
  <div>testValue</div>
  <h3>Prop</h3>
  <div></div>
  <h3>Children</h3>
  <div>
    <div>a</div>
    <div>b</div>
  </div>
</div>"
`;
